src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/Formatting/FormattingTestBase.cs

```diff
@@ -0,0 +1,153 @@
+    {
+        protected async Task RunFormattingTestAsync(string input, string expected, int tabSize = 4, bool insertSpaces = true, string fileKind = null)
+        {
+            // Arrange
+            fileKind ??= FileKinds.Component;
+            var start = input.IndexOf('|', StringComparison.Ordinal);
+            var end = input.LastIndexOf('|');
+            input = input.Replace("|", string.Empty, StringComparison.Ordinal);
+
+            var source = SourceText.From(input);
+            var span = TextSpan.FromBounds(start, end - 1);
+            var range = span.AsRange(source);
+
+            var path = "file:///path/to/document.razor";
+            var uri = new Uri(path);
+            var (codeDocument, documentSnapshot) = CreateCodeDocumentAndSnapshot(source, uri.AbsolutePath, fileKind: fileKind);
+            var options = new FormattingOptions()
+            {
+                TabSize = tabSize,
+                InsertSpaces = insertSpaces,
+            };
+
+            var formattingService = CreateFormattingService(codeDocument);
+
+            // Act
+            var edits = await formattingService.FormatAsync(uri, documentSnapshot, range, options);
+
+            // Assert
+            var edited = ApplyEdits(source, edits);
+            var actual = edited.ToString();
+            Assert.Equal(expected, actual);
+        }
+
+        protected async Task RunOnTypeFormattingTestAsync(string input, string expected, string triggerCharacter, int tabSize = 4, bool insertSpaces = true, string fileKind = null)
+        {
+            // Arrange
+            fileKind ??= FileKinds.Component;
+            var beforeTrigger = input.IndexOf('|', StringComparison.Ordinal);
+            var afterTrigger = input.LastIndexOf('|') - 1;
+            input = input.Replace("|", string.Empty, StringComparison.Ordinal);
+
+            var source = SourceText.From(input);
+
+            var path = "file:///path/to/document.razor";
+            var uri = new Uri(path);
+            var (codeDocument, documentSnapshot) = CreateCodeDocumentAndSnapshot(source, uri.AbsolutePath, fileKind: fileKind);
+            var options = new FormattingOptions()
+            {
+                TabSize = tabSize,
+                InsertSpaces = insertSpaces,
+            };
+
+            var formattingService = CreateFormattingService(codeDocument);
+            var (kind, projectedEdits) = GetFormattedEdits(codeDocument, expected, beforeTrigger);
+
+            // Act
+            var edits = await formattingService.ApplyFormattedEditsAsync(uri, documentSnapshot, kind, projectedEdits, options);
+
+            // Assert
+            var edited = ApplyEdits(source, edits);
+            var actual = edited.ToString();
+            Assert.Equal(expected, actual);
+        }
+
+        private (RazorLanguageKind, TextEdit[]) GetFormattedEdits(RazorCodeDocument codeDocument, string expected, int positionBeforeTriggerChar)
```

Comment:
This is the part where I mimic a real C# and HTML formatter. This is not ideal and needs to be replaced with a more robust infrastructure.